runtime/parachains/src/reward_points.rs

```diff
@@ -23,33 +23,107 @@
 
 use primitives::v1::ValidatorIndex;
 use pallet_staking::SessionInterface;
+use crate::shared;
 
 /// The amount of era points given by backing a candidate that is included.
 pub const BACKING_POINTS: u32 = 20;
 
 /// Rewards validators for participating in parachains with era points in pallet-staking.
 pub struct RewardValidatorsWithEraPoints<C>(sp_std::marker::PhantomData<C>);
 
-fn reward_by_indices<C, I>(points: u32, indices: I) where
-	C: pallet_staking::Config,
+fn validators_to_reward<C, T, I>(validators: &'_ [T], indirect_indices: I) -> impl IntoIterator<Item=&'_ T> where
+	C: shared::Config,
 	I: IntoIterator<Item = ValidatorIndex>
 {
-	// Fetch the validators from the _session_ because sessions are offset from eras
-	// and we are rewarding for behavior in current session.
-	let validators = C::SessionInterface::validators();
-	let rewards = indices.into_iter()
-		.filter_map(|i| validators.get(i.0 as usize).map(|v| v.clone()))
-		.map(|v| (v, points));
-
-	<pallet_staking::Module<C>>::reward_by_ids(rewards);
+	let validator_indirection = <shared::Module<C>>::active_validator_indices();
+
+	indirect_indices.into_iter()
+		.filter_map(move |i| validator_indirection.get(i.0 as usize).map(|v| v.clone()))
+		.filter_map(move |i| validators.get(i.0 as usize))
 }
 
 impl<C> crate::inclusion::RewardValidators for RewardValidatorsWithEraPoints<C>
-	where C: pallet_staking::Config
+	where C: pallet_staking::Config + shared::Config,
 {
-	fn reward_backing(validators: impl IntoIterator<Item=ValidatorIndex>) {
-		reward_by_indices::<C, _>(BACKING_POINTS, validators);
+	fn reward_backing(indirect_indices: impl IntoIterator<Item=ValidatorIndex>) {
+		// Fetch the validators from the _session_ because sessions are offset from eras
+		// and we are rewarding for behavior in current session.
+		let validators = C::SessionInterface::validators();
```

Comment:
just to clarify, this will return an unshuffled list?

do you think it makes sense to make it more type-safe in the future by introducing a `ShuffledVec` type?

Reply:
Yeah, that is an unshuffled list.

We'd need opaque indices for real type-safety on this, but yes, I think that would be useful in the runtime as it would have prevented this exact error.